### ISSUE TITLE
Username shows up differently in the contributions section

### ISSUE THREAD

**Issue by ndattani** [desktop/desktop#6090]
I checked the "known issues page" and also the existing Issues that were already raised, and in both places searched "contributions" and didn't find anything, and the issue didn't come up as a "related issue" when I started typing in my title of this issue.

I pushed a commit from GitHub desktop (Windows 7) and it shows up as a completely different username when I look here: https://github.com/HPQC-LABS/Papers/commits/master

This also makes it look like I made no contributions on November 3rd: https://github.com/ndattani
But I did make the contribution, it was just from GitHub desktop instead of the usual command line submission.

**Comment by j-f1** [comment]
It looks like your email address in that commit has a typo.

**Comment by billygriffin** [comment]
Since it seems like @j-f1 identified the distinction, I'm going to close this out, but if there's an issue with Desktop that likely caused this, please let us know and we can reopen if that's the case.

**Comment by ndattani** [comment]
Hold on @j-f1 and @billygriffin, if it was the wrong email address, why did the password work? Why would I have had permission to push to the repository then? What was the email address used for that commit anyway? How do you see it? I thought this might be a bug with GitHub desktop (I had to give my username and password in order to push that commit, but since it came from GitHub desktop instead of command line, it's showing up as a different user).

**Comment by j-f1** [comment]
I used the API. Here’s a commit with the correct email: https://api.github.com/repos/HPQC-LABS/Papers/commits/5828f166edadb84d31905ffff0fe95c4080fde58
Here’s one with the incorrect email: https://api.github.com/repos/HPQC-LABS/Papers/commits/1f65292ea36376c8f588d4214a7b79d88caf9ab4

One of the tabs in Desktop ‘s preference window should be helpful.

**Comment by steveward** [comment]
@ndattani The email address that Git associates with your commits is only used for the purpose of identifying who made the commit. GitHub uses that email address to attribute those commits to your GitHub account. Git/GitHub Desktop are not using that email address for any type of authentication -- it is purely for attribution. We have a [help article](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/) that explains this further. This is the section to make note of:

>Note: If your commits are linked to another user, that does not mean the user can access your repository. A user can only access a repository you own if you add them as a collaborator or add them to a team that has access to the repository.

You can verify you are using the correct email address by going to the menu in GitHub Desktop and selecting `File` > `Options` > `Git`. I hope that helps explain things!

**Comment by ndattani** [comment]
Thank you @j-f1  and @steveward. Indeed it seems there's a typo, which makes me surprised the repo even let me push to it !! Anyway, how do I fix the typo so that the email address occurs properly in the contributions section? I also wanted to do it for another user who committed a bunch of updates to a repo but his name never came up on the contributions page because he had been using GitHub Desktop instead of the command line. What makes it annoying is that it looks like there's only 4 contributors to the project even though there's actually 5 (and I'm sure it's even worse for him because he contributed some commits that were accepted by the reviewers, but he's not showing up in the contributions section, which means people think he's lying when he says in job interviews that he contributed). See Yudong Cao does not show up: https://github.com/HPQC-LABS/Book_About_Quadratization/graphs/contributors But obviously according to this he did contirbute: https://github.com/HPQC-LABS/Book_About_Quadratization/commit/778db7ab8ee19702d956098ce920d5daef451995

The problem is that he used GitHub Desktop, whereas when he commits from command line it connects to his profile: https://github.com/orgs/HPQC-LABS/people/yudongcao

Can we change it retro-actively so that he gets credit for all the hard work he did?

**Comment by j-f1** [comment]
You could either have him add the email to his account or use `git filter-branch` to rewrite the old commits.

**Comment by steveward** [comment]
> . Indeed it seems there's a typo, which makes me surprised the repo even let me push to it !!

@ndattani Just to clarify -- if a different user is shown as the commit author it does not mean that they have push access to the repository.

> how do I fix the typo so that the email address occurs properly in the contributions section? 

The user will need to open GitHub Desktop and go to `File` > `Options` > `Git` and make sure the email address there matches the email address associated with their GitHub account. The current email addresses associated with an account can be found at https://github.com/settings/emails.

>Can we change it retro-actively so that he gets credit for all the hard work he did?

To retroactively change the email address associated with commits you will need to rewrite the history of the Git repository. We have a help article that walks through how to do that:

https://help.github.com/articles/changing-author-info/

If you have any questions about this process, let me know.

**Comment by ndattani** [comment]
@j-f1: 

> You could either have him add the email to his account or use `git filter-branch` to rewrite the old commits.

Is it that simple? How come @steveward's method is so much more complicated?

I followed @steveward's email because I didn't see j-f1's message until now. Changing the email address in File -> Options -> Git is not enough. You then have to follow the link Steve Ward gave, and re-write the history. Is j-f1 saying that you can do it by:

1. simply adding the correct email address to the account?
2. running the command "git filter-branch" ?


Anyway, I followed Steve Ward's suggestion because I didn't see j-f1's message until now, and this worked for my personal repository that is not so dangerous to play with. However for the Book that I'm writing with several co-authors, I'm worried about using that method because it says: **"Warning: This action is destructive to your repository's history. If you're collaborating on a repository with others, it's considered bad practice to rewrite published history. You should only do this in an emergency."**

What is meant by "destructive to your repository's history" ? Doesn't the script just change the old email address to the new email address?  As for: "If you're collaborating on a repository with others, it's considered bad practice to rewrite published history", what if all we're trying to do is change the email address associated with a commit from 2 months ago?

**Comment by j-f1** [comment]
> Is it that simple? How come @steveward's method is so much more complicated?

I gave a basic overview of what’s needed — @steveward gave the full instructions.

> What is meant by "destructive to your repository's history" ? Doesn't the script just change the old email address to the new email address?

Git works by hashing all the commit’s metadata, including the changed files, timestamp, author name and email, and parent commits to generate the commit SHA. This means that if you change the email address for a commit, all the commits made afterwards will have new SHAs created, making it difficult to merge local changes with the remote.

**Comment by ndattani** [comment]
@j-f1, I see. Is "difficulty to merge local changes with remote" the only problem ??
So, if I get every collaborator to PCP (pull, commit local changes, push) before I revise the history, would there be any other warning not to revise the history?

**Comment by ndattani** [comment]
@steveward , @j-f1 , @billygriffin , I have run the script that was provided in the link by @steveward , in order to add a user to the "contributions" page, who had made contributions but with different email addresses so it didn't show up at all. I had to do it twice because he used two email address, both not associated with his username.

After doing all that, I went back to my local repository and ran "git pull" then "git push" and nothing else. This resulted in two new commits appearing here: https://github.com/HPQC-LABS/Book_About_Quadratization/commits/master

One is a duplicate of the commit titled "4 things:...."  which wasn't there before. Isn't it weird that a new commit (duplication of the last commit) appeared?
The second new commit is the "merge branch" commit.

I find it VERY strange that now the contributors section: https://github.com/HPQC-LABS/Book_About_Quadratization/graphs/contributors

says that I have 252 commits, rather than ~200 commits that I had before. My added characters jumped from 20,000 to 35,000.  What got added? Why is there 52 new commits appearing in my contributions section, even though I didn't do anything other than running the script to change email addresses?

Also strange is that user nchancel now has 28 commits instead of about 14, and tanbur (whose commits would have been all before the commits of yudongcao which is the user I had to correct the email address of) has not got any "new" commits listed, so it looks like he is only the 3rd biggest contributor (by # of commits) whereas he SHOULD be the 2nd biggest.

Do you know what is going on and how I can correct it?

**Comment by j-f1** [comment]
It looks like the `git pull` might have messed things up — when you rewrite history, it’s usually best to make sure you’re up to date before starting, then use `git push --force` to send the rewritten history to GitHub. The best solution for you would probably be to run `git reflog master` to find the previous version of the commit history, check out that commit with `git checkout <sha>; git branch -f master; git checkout master`, redo the commit rewrite, then run `git push -f` to push the new version of history to GitHub.

**Comment by Gahamelas** [comment]
Do I understand correctly that if you set your email (hidden in my case for privacy) correctly in both Github Desktop and web, it doesn't matter the name you assign to yourself to push commits? In any case, is there a way to correct the so they all match under a certain email address?

**Comment by steveward** [comment]
@Gahamelas that is correct -- your user name will show up on GitHub if the email address used for the commit matches an email address associated with your GitHub account. Your full name will still be viewable in the `.patch` file of the commit on GitHub.

> In any case, is there a way to correct the so they all match under a certain email address?

In order to make previous commits use your email address you would need to [rewrite the history](https://help.github.com/en/github/using-git/changing-author-info) of the repository. We don't recommend doing this since it is destructive to your repository. I hope that helps!

**Comment by Gahamelas** [comment]
Actually, I did not want to change my email address, since its the ID private one provided by Github. It's the name I entered locally in my Desktop what was wrong.
